File: apps/robotlab_x_ui/src/models/ServiceProxy.ts
// managed
// Service Proxy interface for use across the project
export interface ServiceProxy {
  id?: string;
  name: string;
  service_meta_id: string;
  status?: string;
  configured?: boolean;
  pid?: number;
  host?: string;
  port?: number;
  created_at?: string;
  started_at?: string;
  stopped_at?: string;
  error?: string;
  service_config?: any;
}

export function createEmptyServiceProxy(): ServiceProxy {
  return {
    id: undefined,
    name: "",
    service_meta_id: "",
    status: "stopped",
    configured: undefined,
    pid: undefined,
    host: undefined,
    port: undefined,
    created_at: undefined,
    started_at: undefined,
    stopped_at: undefined,
    error: undefined,
    service_config: undefined,
  };
}

// Helper function to validate ServiceProxy object
export function isServiceProxy(obj: any): obj is ServiceProxy {
  return obj && typeof obj === 'object';
}

// Helper function to clone ServiceProxy object
export function cloneServiceProxy(item: ServiceProxy): ServiceProxy {
  return {
    ...item,
  };
}

// JSON Schema for Fastify route validation and OpenAPI documentation
export const ServiceProxySchema = {
  "type": "object",
  "properties": {
    "id": {
      "type": "string",
      "description": "Primary key - set to proxy name on create",
      "example": "pycv"
    },
    "name": {
      "type": "string",
      "description": "Unique human-readable name for this running instance",
      "example": "pycv"
    },
    "service_meta_id": {
      "type": "string",
      "description": "Service catalog reference in name@version format",
      "example": "opencv@4.8.0"
    },
    "status": {
      "type": "string",
      "description": "Runtime state - stopped, installing, installed, starting, running, stopping, error",
      "example": "running"
    },
    "configured": {
      "type": "boolean",
      "description": "Whether this proxy instance has been configured via the config wizard"
    },
    "pid": {
      "type": "integer",
      "description": "OS process ID when the service is running"
    },
    "host": {
      "type": "string",
      "description": "Hostname or IP the service is bound to",
      "example": "127.0.0.1"
    },
    "port": {
      "type": "integer",
      "description": "Port the service is listening on",
      "example": 7070
    },
    "created_at": {
      "type": "string",
      "description": "ISO-8601 timestamp when the proxy record was created"
    },
    "started_at": {
      "type": "string",
      "description": "ISO-8601 timestamp of the most recent successful start"
    },
    "stopped_at": {
      "type": "string",
      "description": "ISO-8601 timestamp of the most recent stop or crash"
    },
    "error": {
      "type": "string",
      "description": "Last error message if status is error"
    },
    "service_config": {
      "description": "Per-instance configuration params populated by the config wizard \u2014 mirrors service_config.params for this proxy",
      "type": "object",
      "additionalProperties": true
    }
  },
  "required": [
    "name",
    "service_meta_id"
  ]
}

File: apps/robotlab_x/src/robotlab_x/models/service_proxy.py
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict
from typing import List, Any, Optional, Literal, Dict
import time
import json
import yaml
from database.types import DateTimeStr

class ServiceProxy(BaseModel):
    id: Optional[str] = Field(None, description="Primary key - set to proxy name on create", json_schema_extra={"example":"pycv"})
    name: str = Field(..., description="Unique human-readable name for this running instance", json_schema_extra={"example":"pycv"})
    service_meta_id: str = Field(..., description="Service catalog reference in name@version format", json_schema_extra={"example":"opencv@4.8.0"})
    status: Optional[str] = Field("stopped", description="Runtime state - stopped, installing, installed, starting, running, stopping, error", json_schema_extra={"example":"running"})
    configured: Optional[bool] = Field(None, description="Whether this proxy instance has been configured via the config wizard")
    pid: Optional[int] = Field(None, description="OS process ID when the service is running")
    host: Optional[str] = Field(None, description="Hostname or IP the service is bound to", json_schema_extra={"example":"127.0.0.1"})
    port: Optional[int] = Field(None, description="Port the service is listening on", json_schema_extra={"example":7070})
    created_at: Optional[str] = Field(None, description="ISO-8601 timestamp when the proxy record was created")
    started_at: Optional[str] = Field(None, description="ISO-8601 timestamp of the most recent successful start")
    stopped_at: Optional[str] = Field(None, description="ISO-8601 timestamp of the most recent stop or crash")
    error: Optional[str] = Field(None, description="Last error message if status is error")
    service_config: Optional[dict] = Field(None, description="Per-instance configuration params populated by the config wizard — mirrors service_config.params for this proxy", json_schema_extra={"type":"object","additionalProperties":True})

# managed
    @classmethod
    def from_yaml(cls, yaml_str: str):
        """Load ServiceProxy from a YAML string."""
        data = yaml.safe_load(yaml_str)
        return cls(**data)

    @classmethod
    def from_yaml_file(cls, file_path: str):
        """Load ServiceProxy from a YAML file."""
        with open(file_path, "r") as file:
            return cls.from_yaml(file.read())

    @classmethod
    def from_json(cls, json_str: str):
        """Load ServiceProxy from a JSON string."""
        data = json.loads(json_str)
        return cls(**data)

    @classmethod
    def from_json_file(cls, file_path: str):
        """Load ServiceProxy from a JSON file."""
        with open(file_path, "r") as file:
            return cls.from_json(file.read())
